### ISSUE TITLE
Use dio as default HttpClient

### ISSUE THREAD

**Issue by nioncode** [flutterchina/dio#646]
Is it possible to somehow create an `HttpClient` that wraps all dio functionalities and thus provides e.g. interceptors, retry logic, etc. for all users of `HttpCiient` automatically?

Some widgets in Flutter use an `HttpClient` client automatically (e.g. `NetworkImage`) and can't be told to use dio instead if I'm not mistaken.

It is possible to overwrite the default `HttpClient` globally with the following code:
```dart
class MyHttpOverrides extends HttpOverrides {
  @override
  HttpClient createHttpClient(SecurityContext context) {
    // Return a HttpClient that has all dio features, like
    // retries, interceptors, ...
    // ???
  }
}

void main(){
  HttpOverrides.global = MyHttpOverrides();
  runApp(App());
}
```

Is it possible to create such a wrapper `HttpClient`? This would allow dio to be a drop-in replacement for existing apps to add all kinds of functionality on top of already existing code.

**Comment by shaxxx** [comment]
You wolud need to implement HttpClient abstract class from SDK with your own implementation that uses Dio under the hood. Then just provide your HttpClient to the widget.

But IMHO this is bad idea. Put all your requests behind one "requester" class that handles all requests and retry/interception logic. You don't need to use NetworkImage, use plain image widget, get the image bytes with requester and use it in the image widget.  You can find example here
https://github.com/shaxxx/EnigmaWeb.Dart/blob/master/lib/src/web_requester.dart
Basically, EnigmaWeb is a Dart package that implements API calls on the client side. It's used inside Signalmeter2 app to call web server API. None of the widgets makes network calls without API.
There's even screenshot feature that loads image from server with progress, retry, etc..
https://github.com/shaxxx/Signalmeter2/blob/master/lib/src/ui/screenshot/screenshot_view.dart

Hope this helps.

**Comment by nioncode** [comment]
Thanks for the idea, but many widgets and third party libraries use `HttpClient` and it would be a huge waste of time to re-implement all those dependencies simply for adding dio into them if this could be solved more easily instead.

Since the whole framework and basically the whole community relies on `HttpClient`, I think it would be best to provide an implementation of it in dio so that it could be used as a drop-in replacement instead of having to check each dependency to even know whether they use `HttpClient` under the hood or not.

As dio already uses an `HttpClient` by itself anyways, I think all functionality of dio could also be provided by subclassing `HttpClient` instead of just using it to dispatch the actual http calls.

**Comment by shaxxx** [comment]
Just curious which widgets use HttpClient directly (apart from NetworkImage)?

**Comment by nioncode** [comment]
E.g. https://github.com/mchome/flutter_advanced_networkimage

**Comment by shaxxx** [comment]
This is still an issue

**Comment by shaxxx** [comment]
It's still an issue
